File: TCO 2017 Round 2B/DengklekGaneshAndTree.java
import java.io.*;
import java.util.*;

public class DengklekGaneshAndTree {
    private Vertex[] vs;
    private boolean[] was;
    private Segment dfs(int u) {
        was[u] = true;
        Segment s = new Segment(vs[u].h);
        for (int v : vs[u].edges) {
            if (vs[v].ch == vs[u].ch) {
                s.update(dfs(v));
            }
        }
        return s;
    }
	public int getCount(String labels, int[] parents) {
        int n = labels.length();
        vs = new Vertex[n];
        for (int i = 0; i < n; ++i) {
            vs[i] = new Vertex(labels.charAt(i));
        }
        int high = 0;
        for (int i = 1; i < n; ++i) {
            int pa = parents[i - 1];
            vs[pa].edges.add(i);
            vs[i].h = vs[pa].h + 1;
            high = Math.max(high, vs[i].h);
        }
        ++high;
        was = new boolean[n];
        List<Segment> a = new ArrayList<Segment>();
        for (int i = 0; i < n; ++i) {
            if (!was[i]) {
                a.add(dfs(i));
            }
        }
        Collections.sort(a);
        int[] dp = new int[n+1];
        int[] cur = new int[n+1];
        dp[0] = 1;
        for (Segment s : a) {
            for (int i = 0; i < n+1; ++i) cur[i] = dp[i];
            for (int i = s.a; i <= high; ++i) {
                int nxt = Math.max(i, s.b);
                cur[nxt] = (cur[nxt] + dp[i]) % 1000000007;
            }
            for (int i = 0; i < n+1; ++i) dp[i] = cur[i];
        }

		return dp[high];
	}
    private static class Vertex {
        public int h;
        public char ch;
        public List<Integer> edges = new ArrayList<Integer>();
        public Vertex(){}
        public Vertex(char ch) {
            this.ch = ch;
            this.h = 0;
        }
    }
    private static class Segment implements Comparable {
        public int a;
        public int b;
        public Segment(int a) {
            this.a = a;
            this.b = a + 1;
        }
        public void update(Segment s) {
            a = Math.min(a, s.a);
            b = Math.max(b, s.b);
        }
        public int compareTo(Object o) {
            Segment s = (Segment) o;
            return this.a - s.a;
        }
    }

// CUT begin
	public static void main(String[] args){
		System.err.println("DengklekGaneshAndTree (500 Points)");
		System.err.println();
		HashSet<Integer> cases = new HashSet<Integer>();
        for (int i = 0; i < args.length; ++i) cases.add(Integer.parseInt(args[i]));
        runTest(cases);
	}

	static void runTest(HashSet<Integer> caseSet) {
	    int cases = 0, passed = 0;
	    while (true) {
	    	String label = Reader.nextLine();
	    	if (label == null || !label.startsWith("--"))
	    		break;

            String labels = Reader.nextLine();
            int[] parents = new int[Integer.parseInt(Reader.nextLine())];
            for (int i = 0; i < parents.length; ++i)
                parents[i] = Integer.parseInt(Reader.nextLine());
            Reader.nextLine();
            int __answer = Integer.parseInt(Reader.nextLine());

            cases++;
            if (caseSet.size() > 0 && !caseSet.contains(cases - 1))
                continue;
    		System.err.print(String.format("  Testcase #%d ... ", cases - 1));

            if (doTest(labels, parents, __answer))
                passed++;
	    }
	    if (caseSet.size() > 0) cases = caseSet.size();
        System.err.println(String.format("%nPassed : %d/%d cases", passed, cases));

        int T = (int)(System.currentTimeMillis() / 1000) - 1500710869;
        double PT = T / 60.0, TT = 75.0;
        System.err.println(String.format("Time   : %d minutes %d secs%nScore  : %.2f points", T / 60, T % 60, 500 * (0.3 + (0.7 * TT * TT) / (10.0 * PT * PT + TT * TT))));
	}

	static boolean doTest(String labels, int[] parents, int __expected) {
        labels = new String(labels);
		long startTime = System.currentTimeMillis();
		Throwable exception = null;
		DengklekGaneshAndTree instance = new DengklekGaneshAndTree();
		int __result = 0;
		try {
			__result = instance.getCount(labels, parents);
		}
		catch (Throwable e) { exception = e; }
		double elapsed = (System.currentTimeMillis() - startTime) / 1000.0;

		if (exception != null) {
			System.err.println("RUNTIME ERROR!");
			exception.printStackTrace();
			return false;
		}
		else if (__result == __expected) {
			System.err.println("PASSED! " + String.format("(%.2f seconds)", elapsed));
			return true;
		}
		else {
			System.err.println("FAILED! " + String.format("(%.2f seconds)", elapsed));
			System.err.println("           Expected: " + __expected);
			System.err.println("           Received: " + __result);
			return false;
		}
	}

	static class Reader {
        private static final String dataFileName = "DengklekGaneshAndTree.sample";
	    private static BufferedReader reader;

	    public static String nextLine() {
	        try {
                if (reader == null) {
                    reader = new BufferedReader(new InputStreamReader(new FileInputStream(dataFileName)));
                }
                return reader.readLine();
	        }
	        catch (IOException e) {
	            System.err.println("FATAL!! IOException");
	            e.printStackTrace();
	            System.exit(1);
	        }
	        return "";
	    }
	}
// CUT end
}

File: TCO 2017 Round 2B/DengklekGaneshAndTree.cpp
#include <cstdio>
#include <cmath>
#include <cstring>
#include <ctime>
#include <iostream>
#include <algorithm>
#include <set>
#include <vector>
#include <sstream>
#include <typeinfo>
#include <fstream>

using namespace std;
const int MOD = 1e9 + 7;
struct Vertex {
    int h;
    char c;
    vector<int> edges;
};
struct Segment {
    int a, b;
    Segment(int a = 0) : a(a), b(a + 1) {}
    void add(const Segment &s) {
        a = min(a, s.a);
        b = max(b, s.b);
    }
    bool operator < (const Segment &s) const {
        return a < s.a;
    }
};

class DengklekGaneshAndTree {
    public:
    vector<Vertex> vs;
    vector<bool> was;
    Segment dfs(int u) {
        was[u] = true;
        Segment cur(vs[u].h);
        for (int v : vs[u].edges) {
            if (vs[u].c == vs[v].c) {
                cur.add(dfs(v));
            }
        }
        return cur;
    }
    int getCount(string labels, vector<int> parents) {
        int n = labels.size();
        vs.resize(n);
        for (int i = 0; i < n; ++i) {
            vs[i].c = labels[i];
        }
        int high = 0;
        for (int i = 1; i < n; ++i) {
            int pa = parents[i -1];
            vs[pa].edges.push_back(i);
            vs[i].h = vs[pa].h + 1;
            high = max(high, vs[i].h);
        }
        high = high + 1;
        vector<Segment> a;
        was.assign(n, false);
        for (int i = 0; i < n; ++i) {
            if (!was[i]) {
                a.push_back(dfs(i));
            }
        }
        sort(a.begin(), a.end());
        vector<int> dp(high + 1, 0);
        dp[0] = 1;
        for (Segment &s : a) {
            vector<int> cur(dp);
            for (int i = s.a; i <= high; ++i) {
                cur[max(s.b, i)] = (cur[max(s.b, i)] + dp[i]) % MOD;
            }
            dp = cur;
        }

        return dp[high];
    }
};

// CUT begin
ifstream data("DengklekGaneshAndTree.sample");

string next_line() {
    string s;
    getline(data, s);
    return s;
}

template <typename T> void from_stream(T &t) {
    stringstream ss(next_line());
    ss >> t;
}

void from_stream(string &s) {
    s = next_line();
}

template <typename T> void from_stream(vector<T> &ts) {
    int len;
    from_stream(len);
    ts.clear();
    for (int i = 0; i < len; ++i) {
        T t;
        from_stream(t);
        ts.push_back(t);
    }
}

template <typename T>
string to_string(T t) {
    stringstream s;
    s << t;
    return s.str();
}

string to_string(string t) {
    return "\"" + t + "\"";
}

bool do_test(string labels, vector<int> parents, int __expected) {
    time_t startClock = clock();
    DengklekGaneshAndTree *instance = new DengklekGaneshAndTree();
    int __result = instance->getCount(labels, parents);
    double elapsed = (double)(clock() - startClock) / CLOCKS_PER_SEC;
    delete instance;

    if (__result == __expected) {
        cout << "PASSED!" << " (" << elapsed << " seconds)" << endl;
        return true;
    }
    else {
        cout << "FAILED!" << " (" << elapsed << " seconds)" << endl;
        cout << "           Expected: " << to_string(__expected) << endl;
        cout << "           Received: " << to_string(__result) << endl;
        return false;
    }
}

int run_test(bool mainProcess, const set<int> &case_set, const string command) {
    int cases = 0, passed = 0;
    while (true) {
        if (next_line().find("--") != 0)
            break;
        string labels;
        from_stream(labels);
        vector<int> parents;
        from_stream(parents);
        next_line();
        int __answer;
        from_stream(__answer);

        cases++;
        if (case_set.size() > 0 && case_set.find(cases - 1) == case_set.end())
            continue;

        cout << "  Testcase #" << cases - 1 << " ... ";
        if ( do_test(labels, parents, __answer)) {
            passed++;
        }
    }
    if (mainProcess) {
        cout << endl << "Passed : " << passed << "/" << cases << " cases" << endl;
        int T = time(NULL) - 1500707808;
        double PT = T / 60.0, TT = 75.0;
        cout << "Time   : " << T / 60 << " minutes " << T % 60 << " secs" << endl;
        cout << "Score  : " << 500 * (0.3 + (0.7 * TT * TT) / (10.0 * PT * PT + TT * TT)) << " points" << endl;
    }
    return 0;
}

int main(int argc, char *argv[]) {
    cout.setf(ios::fixed, ios::floatfield);
    cout.precision(2);
    set<int> cases;
    bool mainProcess = true;
    for (int i = 1; i < argc; ++i) {
        if ( string(argv[i]) == "-") {
            mainProcess = false;
        } else {
            cases.insert(atoi(argv[i]));
        }
    }
    if (mainProcess) {
        cout << "DengklekGaneshAndTree (500 Points)" << endl << endl;
    }
    return run_test(mainProcess, cases, argv[0]);
}
// CUT end
